### PR TITLE
Create IP to hostname mapping for Medusa restores

### LINKED ISSUE #486: K8SSAND-1357 ⁃ Create Medusa restore mapping in the operator
Depending on the results of #485, we should be able to compute restore mappings in the operator directly.
After getting the backup tokenmap from the storage backend, we can generate a mapping file in a config map that will be mounted on the pods for Medusa to read it.
The mapping should look like the following:

```
{
  'in_place': false,
  'host_map':
  {
     'target-node-0': 'source-node-1',
     'target-node-1': 'source-node-3',
     'target-node-2': 'source-node-0',
    ...
  }
}
```

The mapping should respect rack assignments and detecting whether restore is in place or not can be based on overlaps between source node names and projected target pod names.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-1357) by [Unito](https://www.unito.io)
┆friendlyId: K8SSAND-1357
┆priority: Medium

### PR DESCRIPTION
**Pull request by Miles-Garnsey**
**What this PR does**:

Creates a mapping from the hosts in a given remote Medusa backup (on a cloud bucket) to targets derived from a K8ssandraCluster's projected hostnames.

**Which issue(s) this PR fixes**:
Fixes #486 

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Review by adejanovski — COMMENTED** [review]
Great stuff so far @Miles-Garnsey !

I left a few minor comments, but this is clearly headed in the right direction.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:92]: > You may want to trade "IPs" for "Host" in the following functions since we won't be dealing with just IPs. We can have hostnames as well.

Agreed and done. This fell out of sync when I was making changes to accomodate IPs in the source.

> Also the comment repeats (incorrectly) the name of the function, which is probably not something that's desirable: if you change the name of the function (as you did) then you need to change it in the comment as well.

It is indeed unfortunate, however it is [required](https://go.dev/blog/godoc) for godoc to pick up the comment as documenting the function:

> to document a type, variable, constant, function, or even a package, write a regular comment directly preceding its declaration, with no intervening blank line. 

Whether we strictly need to document a private function I'm not sure, but I'm trying to make a habit of documenting as much as possible.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:147]: Sure, done.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:174]: As above, Godoc requirement is to repeat the function name. 

> Also, we shouldn't mention "IP" here either since we can get hostnames as sources.

That's valid, I'll change it.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:175]: This is actually all handled now, so no need for the TODO.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:186]: There are three solutions to this problem.

### Solution 1

In the current solution, rack names must match, and the racks must be the same size between the source and destination topologies, but they do **not** need to be balanced (i.e. all racks the same size). Additionally, users can control the mapping with precision.

### Solution 2

Under your proposal, rack names may differ, but the racks must be balanced and users cannot control the placements with precision.

### Solution 3

If we don't care about being able to control placement, we can eliminate requirements on both balance and naming via the following:
1. Add a function which takes a `map[nodeLocation]string` and returns `map[len(nodes)]nodeLocation`.
2. Put sources and targets though that function.
3. Do a "join" on the two maps via a for loop using len(nodes) as the key.

This has the advantage that racks may be unbalanced AND names need not match. But users still cannot control placement since (in the standard case where all racks are the same size) the source-destination mapping will be randomised.

In some ways, I'd rather force the rack names to match as it gives more control. But I'm not committed to any particular solution here.

File: pkg/medusa/hostmap.go
Comment (adejanovski) [reply pkg/medusa/hostmap.go:155]: It's fine to use the K8ssandraCluster spec for now. We don't have the full design of the restore on init feature so we can revisit this in the future if needed.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go]: Discussion we had offline (with thanks to @adejanovski for the info):

* Network topology strategy actually allocates replicas based on rack names (replicas go to the racks alphabetically close to the primary replica).
* The EC2 aware snitch will actually modify rack names.

As a result of these two facts, it isn't appropriate to (a) force users to use the same rack names between source and destination clusters or (b) allow customers to change the token ordering relative to the rack-name alphabetical ordering.

Further changes to account for this logic incoming.

Comment (jsanda) [reply pkg/medusa/hostmap.go]: >The EC2 aware snitch will actually modify rack names

Does it matter that we always use GossipingPropertyFileSnitch?

Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go]: I'm not sure @jsanda but it is a worthwhile callout. Based on conversation with Alexander, it sounded like he was also concerned about the replicas being allocated into the same racks (which happens based on the alphabetical ordering of their names). I didn't quite follow that part 100%.

@adejanovski your thoughts are welcome. I have to be honest, the most recent round of changes has left me less than happy with this code as I feel that the data structures etc. that I originally wrote (when I was going to force rack names to match) are now poorly suited to the new requirements around alphabetical ordering.

Comment (adejanovski) [reply pkg/medusa/hostmap.go]: > Does it matter that we always use GossipingPropertyFileSnitch?

No, it's fine. It's just worth mentioning the EC2Snitch because the source cluster could be using it.

**Comment by adejanovski** [comment]
It all looks very good to me @Miles-Garnsey 👍 

One last thing is missing if you check #486, which is the `in_place` setting.
If the mapping shows exactly matching pod names between the source and the target cluster, then you're performing an in place restore (and `in_place` should then be `true`), otherwise it's a remote restore (in this case we don't restore the `system.local` table which contain the cluster name and tokens) .

**Comment by Miles-Garnsey** [comment]
@adejanovski my latest commit adds a function `IsInPlace()` to check for the condition you've described. Let me know if it is what you need.

**Comment by Miles-Garnsey** [comment]
Final unit tests finished/added. I also noticed that there is no way that we can actually restore a cluster with unbalanced racks, because we don't allow for setting rack size anywhere at the cass-operator level [here](https://github.com/k8ssandra/cass-operator/blob/master/apis/cassandra/v1beta1/cassandradatacenter_types.go#L303) so I guess all of those considerations weren't actually worth worrying about too much (or would be part of a different ticket anyway).

I think this is ready to merge if there isn't anything else to consider.

**Review by adejanovski — APPROVED** [review]
Just a tiny request for one or two additional unit tests, but otherwise it all looks very good to me 👍 

Let's hold off from merging at the moment as this PR relies on another one that's yet to be approved.
This work may need some rebasing if the solution for spinning up the standalone medusa pod turns out to be different than what's currently implemented.

File: pkg/medusa/hostmap.go
Comment (Miles-Garnsey) [reply pkg/medusa/hostmap.go:244]: Gah, sorry, I did have these written, I just hadn't committed/pushed. Don't know how I missed that. Check again now.

**Comment by adejanovski** [comment]
Closing in favor or #980

**Pull request closed without merging**
